kroxylicious/src/main/java/io/kroxylicious/proxy/internal/filter/UpstreamBrokerAddressCachingNetFilter.java

```diff
@@ -30,28 +29,25 @@ public class UpstreamBrokerAddressCachingNetFilter implements NetFilter {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(UpstreamBrokerAddressCachingNetFilter.class);
 
-    private final HostPort targetClusterBootstrap;
-    private final FilterChainFactory filterChainFactory;
-    private final ClusterEndpointConfigProvider endpointConfigProvider;
+    private final Configuration config;
+    private final VirtualCluster virtualCluster;
 
-    private final Map<Integer, HostPort> upstreamBrokers = new ConcurrentHashMap<>();
-
-    public UpstreamBrokerAddressCachingNetFilter(HostPort targetClusterBootstrap, FilterChainFactory filterChainFactory,
-                                                 ClusterEndpointConfigProvider endpointConfigProvider) {
-        this.targetClusterBootstrap = targetClusterBootstrap;
-        this.filterChainFactory = filterChainFactory;
-        this.endpointConfigProvider = endpointConfigProvider;
+    public UpstreamBrokerAddressCachingNetFilter(Configuration config, VirtualCluster virtualCluster) {
+        this.config = config;
+        this.virtualCluster = virtualCluster;
     }
 
     @Override
     public void selectServer(NetFilterContext context) {
+        var filterChainFactory = new FilterChainFactory(config, virtualCluster.getClusterEndpointProvider());
+
         var filters = new ArrayList<>(Arrays.stream(filterChainFactory.createFilters()).toList());
 
         // Add a filter to the *end of the chain* that gathers the true nodeId/upstream broker mapping.
         filters.add((MetadataResponseFilter) (header, response, filterContext) -> {
             response.brokers().forEach(b -> {
                 var replacement = new HostPort(b.host(), b.port());
-                var existing = upstreamBrokers.put(b.nodeId(), replacement);
+                var existing = virtualCluster.getUpstreamClusterCache().put(b.nodeId(), replacement);
```

Comment:
I know this is an interim state but would prefer not to expose `getUpstreamClusterCache` from the `virtualCluster` bot to have the virtualCluster expose `cacheNodeMaping(b.nodeId(), replacement)`